JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,489 @@
+// bad
+<ChildComponent
+    onClick={this.onClickHandler}
+    isPropActive={true}
+/>
+
+// good
+<ChildComponent
+    onClick={this.onClickHandler}
+    isPropActive
+/>
+```
+
+## Receiving Props
+- always use type checking for props
+- use `defaultProps` if you need to set default values in cases when the prop value isn't received
+
+``` javascript
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+
+class ComponentName extends Component {
+    // ...
+    render() {
+        return (
+          <div>{this.props.name}</div>
+        );
+      }
+}
+
+ComponentName.propTypes = {
+    id: PropTypes.string.isRequired,
+    name: PropTypes.string,
+};
+
+ComponentName.defaultProps = {
+    name: 'Default',
+};
+```
+
+- you can also declare property types as a **static member** inside a class:
+
+``` javascript
+class ComponentName extends Component {
+    static propTypes = {
+        id: PropTypes.string.isRequired,
+        name: PropTypes.string,
+    }
+    // ...
+    render() {
+        return (
+          <div>{this.props.name}</div>
+        );
+      }
+}
+```
+
+- inside the *render* method, preffer accessing the props using **object destructuring**:
+
+``` javascript
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+
+class ComponentName extends Component {
+    // ...
+    render() {
+        const {id, name} = this.props;
+        // ...
+        return (
+          <div>{name}</div>
+        );
+    }
+}
+```
+
+- when using TypeScript with React, use built in types instead of React's `prop-types` package
+- add `?` at the end of parameters that are optional
+
+``` javascript
+class componentName extends Component<{
+  className?: string;
+  isConditionMet: boolean;
+  onClick?: (event: any) => void;
+}, {}> {
+  // ...
+}
+```
+
+- if you want a cleaner component declaration, or if a component receives a great number of props, use an interface
+- use **"I"** prefixes in interface names
+
+
+``` javascript
+interface IComponentProps {
+  className?: string;
+  isConditionMet: boolean;
+  onClick?: (event: any) => void;
+}
+
+class ComponentName extends Component<IComponentProps, {}> {
+  // ...
+}
+
+// in pure function components
+interface IComponentProps {
+  className?: string;
+  isConditionMet: boolean;
+  onClick?: (event: any) => void;
+}
+
+function ComponentName(props: IComponentProps) {
+  return (
+    // ...
+  );
+}
+```
+
+
+## Conditional Rendering
+- use ternary operators for conditional rendering
+- return `null` if the condition is not met
+
+``` javascript
+render() {
+    // ...
+    return (
+        <AlwaysShownComponent />
+        
+        {
+            isConditionMet
+            ? <ConditionallyShownComponent />
+            : null;
+        }
+    );
+}
+```
+
+## Rendering Arrays of Data
+- use array's **map** method to render items from an array
+- you should have a unique key for each element
+- **don't use array's index as a key**
```

Comment:
there is a use case, mention when it's a thing and when it's not. what about local arrays?